### ISSUE TITLE
Should `as_function()` have  `arg_name`?

### ISSUE THREAD

**Issue by hadley** [r-lib/rlang#1239]
Since `as_function()` is often used for type coercion/checking on input arguments, and it would be nice to include the argument name in the output. Might be worth considering as part of the general problem of getting more context into errors, e.g.:

```R
foo <- function(f) {
  f <- as_function(f)
}
foo(1)
#> Error in foo(): `f` must be a function, not a double vector
```

Also related to the type annotations that @jimhester will be working on. We can probably give the low-level rlang functions an explicit interface with the understanding that that code will eventually be automatically generated.

```R
foo <- function(f) {
  f <- as_function(f, "foo()", "f()`)
  f(1)
}

# Would eventually become something like this
#' @type f function
foo <- function(f) {
}
```

**Comment by lionel-** [comment]
Does this mean all input checking functions should take two arguments like this?

```r
error_call = sys.call(sys.parent(2)),
error_arg = substitute(x)
```

**Comment by hadley** [comment]
Yeah, that's what I'm wondering. 

Also not sure if better to supply a call or a frame — call is more flexible, but frame is slightly easier to pass through and we have a lot of existing functions that do `env = caller_env()`.

**Comment by lionel-** [comment]
Yeah I was thinking the same, especially for the C-level interface. To that end I'm accepting `call = <env>` in the branch I'm working on, this way we don't need two parameters and their interaction.
